pkg/client/client_v2.go

```diff
@@ -581,6 +602,80 @@ func (c *clientV2) tryPerformAction(actionResults chan *proto.ActionResponse, ac
 	}()
 }
 
+func (c *clientV2) tryPerformDiagnostics(actionResults chan *proto.ActionResponse, action *proto.ActionRequest) {
+	// find the unit
+	c.unitsMu.RLock()
+	unit := c.findUnit(action.UnitId, UnitType(action.UnitType))
+	c.unitsMu.RUnlock()
+	if unit == nil {
+		actionResults <- &proto.ActionResponse{
+			Token:  c.token,
+			Id:     action.Id,
+			Status: proto.ActionResponse_FAILED,
+			Result: ActionErrUnitNotFound,
+		}
+		return
+	}
+
+	// gather diagnostics hooks for this unit
+	diagHooks := make(map[string]diagHook)
+	c.dmx.RLock()
+	for n, d := range c.diagHooks {
+		diagHooks[n] = d
+	}
+	c.dmx.RUnlock()
+
+	// unit hooks after client hooks; allows unit specific to override client hooks
+	unit.dmx.RLock()
+	for n, d := range unit.diagHooks {
```

Comment:
Is there a reason not to do this in the same lock protected block as `for n, d := range c.diagHooks` above? I assume it is intentionally the same mutex for both maps?

It's not incorrect I just wanted to double check the intent here.

Reply:
Different lock is used because each unit can have its own hooks and can be registered at any time. So we need to have a unique lock per unit for hook registration.